### PR TITLE
Experimental zip export for ONNX 

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def _trace(func, args, return_outs=False, aten=False):
 
 
 def _export(model, args, f, export_params=True, verbose=False, training=False,
-            input_names=None, output_names=None, aten=False):
+            input_names=None, output_names=None, aten=False, export_zip=False, compress_zip=False):
```
Comment (dzhulgakov) [inline torch/onnx/utils.py:117]: have enum or something for format?

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -147,11 +147,20 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
...
+        proto, export_map = trace.export(list(_unique_state_dict(model).values()), _onnx_opset_version, export_zip)
     else:
-        proto = trace.export([], _onnx_opset_version)
+        proto, export_map = trace.export([], _onnx_opset_version)
+
+    if export_zip:
+        import zipfile
+        compression = zipfile.ZIP_DEFLATED if compress_zip else zipfile.ZIP_STORED
+        with zipfile.ZipFile(f, 'w', compression=compression) as z:
+            z.writestr('__GRAPH_PROTO', proto)
```
Comment (dzhulgakov) [inline torch/onnx/utils.py:158]: constant for it? and it's model proto btw

Comment (dzhulgakov) [reply torch/onnx/utils.py:158]: another crazy ideas is that zip files can have arbitrary crap in the beginning of the file:

```
$ unzip -l foo.zip
Archive:  foo.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        6  03-23-2018 15:48   a
        7  03-23-2018 15:48   b
---------                     -------
       13                     2 files

$ (echo "crap, this is random crap" && cat foo.zip) > bar.zip

unzip -l bar.zip
Archive:  bar.zip
warning [bar.zip]:  26 extra bytes at beginning or within zipfile
  (attempting to process anyway)
  Length      Date    Time    Name
---------  ---------- -----   ----
        6  03-23-2018 15:48   a
        7  03-23-2018 15:48   b
---------                     -------
       13                     2 files

$ ls -la *.zip
-rw-r--r--. 1 dzhulgakov users 321 Mar 23 15:51 bar.zip
-rw-r--r--. 1 dzhulgakov users 295 Mar 23 15:49 foo.zip
```

So we could just prepend the onnx proto as the first thing so deserializing as proto message directly still works. Not sure it's a good idea though :)

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -147,11 +147,20 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
...
+            for k, v in export_map.items():
+                z.writestr(k, v)
+    else:
+        torch.serialization._with_file_like(f, "wb", lambda f: f.write(proto))
```
Comment (dzhulgakov) [inline torch/onnx/utils.py:162]: what was this wrapper achieving?

**Review by dzhulgakov — COMMENTED** [review]
cool! can we avoid clang-formatting the generated files somehow?

File: torch/onnx/utils.py (hunk 2 above)
Comment (jamesr66a) [reply torch/onnx/utils.py:158]: Hmm, that's intriguing. Do you think we would ever end up with multiple ModelProto's in an archive? that would be the only thing stopping me from doing this

**Pull request closed without merging**